### import all the necessary libraries

In [1]:
import json
from collections import defaultdict
from datetime import datetime
import mstarpy


### loading the data from provided json file

In [2]:
def load_transaction_data(file_path):
    with open(file_path) as f:
        data = json.load(f)
    return data['data'][0]['dtSummary']

### Processing the transaction to create a profolio structure

In [3]:
def process_transactions(transactions):
    portfolio = {}
    
    for trxn in transactions:
        scheme = trxn['schemeName']
        folio = scheme  # Assuming folio is represented by scheme name (since not explicitly provided)
        
        units = float(trxn.get('closingBalance', 0))  # Units held
        nav = float(trxn.get('nav', 0))  # Current NAV
        cost_value = float(trxn.get('costValue', 0))  # Initial cost value
        
        if folio not in portfolio:
            portfolio[folio] = {}
        
        portfolio[folio][scheme] = {
            'units': units,
            'nav': nav,
            'cost_value': cost_value
        }
    
    return portfolio


### Function to calculate the gain ang value of the portfolio

In [4]:
def calculate_portfolio_value_and_gain(portfolio):
    total_value = 0
    total_cost = 0 
    for folio, scheme_data in portfolio.items():
        for scheme, data in scheme_data.items():
            current_nav = data['nav']
            units_held = data['units']
            purchase_cost = data['cost_value']
            if units_held > 0 and current_nav > 0:
                current_value = units_held * current_nav
                total_value += current_value
                total_cost += purchase_cost
    total_gain = total_value - total_cost
    return total_value, total_gain

In [8]:
if __name__ == "__main__":
    transactions = load_transaction_data('transaction.json') 
    portfolio = process_transactions(transactions)
    total_value, total_gain = calculate_portfolio_value_and_gain(portfolio)
    print(f"Total Portfolio Value: {total_value}")
    print(f"Total Portfolio Gain: {total_gain}")

Total Portfolio Value: 4089696.7902969
Total Portfolio Gain: 964793.8802968999


### The final Portfolio Value is "4089696.7902969"

### The final Portfolio Gain is "964793.8802968999"